iio: adc: Add initial driver support for MAX14001/MAX14002

MarileneGarcia opened the pull request:
The MAX14001/MAX14002 are configurable, isolated 10-bit ADCs for multi-range binary inputs.

Datasheet:
Link: https://www.analog.com/media/en/technical-documentation/data-sheets/MAX14001-MAX14002.pdf

## PR Description
Basic IIO ADC max14001 driver features:
- Definitions for register addresses and useful constants (if any).
- Definitions for the IIO channels.
- Device state struct.
- read_raw and write_raw function implementations (just a mock for now).
- iio_info struct populated with a reference to the read_raw and write_raw functions.
- Probe function to initialize and make basic setup for the device.
- SPI and device tree MODULE_DEVICE_TABLEs
- Device tree overlay for Raspberry Pi 5
- Basic max14001 SPI read and write functions

## PR Type
- New feature (a change that adds new functionality)

## PR Checklist
- I have conducted a self-review of my own code changes
- I tested it on the Raspberry Pi, but without the max14001 board, I couldn't fully validate it yet.
- Evidence of some basic features working
<img width="1382" height="974" alt="Screenshot from 2025-07-13 20-02-34" src="https://github.com/user-attachments/assets/ac3b5349-47cc-4477-9850-959181cfafcc" />

machschmitt reviewed (commented):
Hi @MarileneGarcia , a few comments about this initial driver version.
Note that, aside from the comments above, there are still some improvements that need to be made to make this driver usable (e.g. read ADC sample data, provide scale to convert output codes to mV). Those features shall be implemented next. Though, overall,  this looks good to me as an initial driver version.

jonathanns reviewed (commented):
Just a few more notes.

MarileneGarcia commented:
> Hi @MarileneGarcia , a few comments about this initial driver version. Note that, aside from the comments above, there are still some improvements that need to be made to make this driver usable (e.g. read ADC sample data, provide scale to convert output codes to mV). Those features shall be implemented next. Though, overall, this looks good to me as an initial driver version.

Ok, thank you for the comments! I think I have address them. I will start working in the other needed features, like the read ADC sample data and provide scale to convert output codes to mV.

MarileneGarcia commented on arch/arm/boot/dts/overlays/rpi-max14001-overlay.dts:
```diff
@@ -0,0 +1,35 @@
+		compatible = "max14001";
+		reg = <0>;
+		spi-max-frequency = <5000000>;
+		status = "okay";
```
Done, thanks.

MarileneGarcia commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,268 @@
+	u16 reversed = 0;
+	int ret = 0;
+
+	pr_err("[Log Debug] max14001_spi_read: reg: %x, val: %x\n", reg, *val);
```
Done, thanks.

MarileneGarcia commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,268 @@
+	struct spi_transfer xfer = {
+		.tx_buf = NULL,
+		.len = 0,
+	};
```
Done, thanks!

MarileneGarcia commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,268 @@
+		.type = IIO_VOLTAGE,
+		.indexed = 1,
+		.channel = 0,
+		.output = 0,
```
Done, thanks!

MarileneGarcia commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,268 @@
+		.type = IIO_CURRENT,
+		.indexed = 1,
+		.channel = 0,
+		.output = 0,
```
Done, thanks!

MarileneGarcia commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,268 @@
+
+static const struct iio_info max14001_info = {
+	.read_raw = max14001_read_raw,
+	.write_raw = max14001_write_raw,
```
Ok, thanks!
I left a comment to check it in the next steps, because I think I may have to use an write attribute.

MarileneGarcia commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,268 @@
+
+MODULE_AUTHOR("Marilene Andrade Garcia <marilene.agarcia@gmail.com>");
+MODULE_DESCRIPTION("Analog Devices MAX14001/MAX14002 ADCs driver");
+MODULE_LICENSE("GPL v2");
```
Done, thanks!

machschmitt commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,305 @@
+{
+	int ret;
+
+	//Enable register write
```
Use IIO subsystem preferred comment style:
```diff
-	//Enable register write
+	/* Enable register write */
```

machschmitt commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,305 @@
+	if (ret < 0)
+		return ret;
+
+	//Write data into register
```
Same here
```diff
-	//Write data into register
+	/* Write data into register */
```

MarileneGarcia replied:
Done, thanks.

machschmitt commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,305 @@
+	if (ret < 0)
+		return ret;
+
+	//Disable register write
```
and here

MarileneGarcia replied:
Done, thanks.

machschmitt commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,305 @@
+	indio_dev->name = st->chip_info->name;
+	indio_dev->modes = INDIO_DIRECT_MODE;
+	indio_dev->info = &max14001_info;
+
```
Missed it on my first review, but the driver needs to get the voltage regulators during probe.
IIRC, vdd and vddl are required while vrefin is not. If so, I'd suggest using `devm_regulator_bulk_get_enable()` to get vdd and vddl, and `devm_regulator_get_enable_read_voltage()` for vrefin. See `drivers/regulator/devres.c`, https://git.kernel.org/pub/scm/linux/kernel/git/jic23/iio.git/tree/drivers/iio/adc/ad7191.c#n130 , for example.

PS.: those regulator_get_... calls may not be available in kernel 6.6. If so, you can either (1) use `devm_regulator_get()`, `regulator_enable()`, and `regulator_get_voltage()`, or (2) update to a newer kernel and look for that API. We have a newer rpi branch now, https://github.com/analogdevicesinc/linux/tree/staging-rpi/rpi-6.12.y

Slight preference for (2), but option (1) is also fine for now.

MarileneGarcia replied:
Ok, thanks! The _devm_regulator_get_enable()_ and _devm_regulator_get_enable_read_voltage()_ were available in my current branch.

machschmitt commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,305 @@
+
+	/* TODO: Validate this line in the hw, could be le16_to_cpu */
+	reversed = bitrev16(be16_to_cpu(rx));
+	*val = FIELD_GET(MAX14001_MASK_ADDR, reversed);
```
why MAX14001_MASK_ADDR?
Isn't the intent to read the register data?

MarileneGarcia replied:
Done, thanks.

machschmitt reviewed (approved):
Looks good to me.

A few more comments.
The regulator thing can be done on a different PR (we might need to rebase or create another GSoC branch if updating to kernel 6.12).

Keep up the good work.

machschmitt commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,359 @@
+	struct device *dev = &st->spi->dev;
+	int ret = 0;
+
+	ret = devm_regulator_get_enable(dev, "vdd");
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to enable specified Vdd supply\n");
+
+	ret = devm_regulator_get_enable(dev, "vddl");
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to enable specified Vddl supply\n");
```
All this could go directly into the probe function. Before reading of/device tree properties. These are required power supplies, so best to fail early if any of them is absent.

MarileneGarcia replied:
Done, thank you!

machschmitt commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,359 @@
+	struct device *dev = &st->spi->dev;
+	int ret = 0;
+
+	ret = devm_regulator_get_enable_read_voltage(dev, "vrefin");
+	if (ret < 0){
+		st->vref_mv = 1250000 / 1000;
+		dev_info(&st->spi->dev, "%s: vrefin not found. vref_mv %d\n", __func__, st->vref_mv);
+	} else {
+		st->vref_mv = ret / 1000;
+		dev_info(&st->spi->dev, "%s: vrefin found. vref_mv %d\n", __func__, st->vref_mv);
+	}
```
This also could go directly in the probe function after we read to required regulators.

MarileneGarcia replied:
Done, thanks!

machschmitt commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,359 @@
+MODULE_AUTHOR("Marilene Andrade Garcia <marilene.agarcia@gmail.com>");
+MODULE_DESCRIPTION("Analog Devices MAX14001/MAX14002 ADCs driver");
+MODULE_LICENSE("GPL v2");
+
```
drop blank line at end of driver file.

MarileneGarcia replied:
Done, thanks!

machschmitt reviewed (approved):
Hi @MarileneGarcia , the updates to MAX14001 driver look mostly good to me. I would merge the PR, but I think it will actually be easier to do the suggested changes before merging this rather than doing them later. So, I'll wait until the last adjustments are done to merge this PR.

MarileneGarcia commented:
> Hi @MarileneGarcia , the updates to MAX14001 driver look mostly good to me. I would merge the PR, but I think it will actually be easier to do the suggested changes before merging this rather than doing them later. So, I'll wait until the last adjustments are done to merge this PR.

Hello @machschmitt. Sure, I have done the requested changes, thank you!
I also submitted the initial version of the device tree documentation. And a new method to set the value of the verification registers, which should be done every time the device powers up.

machschmitt reviewed (changes requested):
Hi @MarileneGarcia , a few comments about the dt-bindings and new code. I'll have a closer look at MAX14001 data sheets to provide a more accurate review this weekend.

machschmitt commented on Documentation/devicetree/bindings/iio/adc/adi,max14001.yaml:
```diff
@@ -0,0 +1,68 @@
+  vrefin-supply:
+    description:
+      Reference input range.
+
```
dt maintainers will probably ask to also document interrupts (even if they are not supported by the Linux driver).

For the interrupt generated by threshold trespassing, I suggest the following.
```yaml
  interrupts:
    maxItems: 2
    items:
      - description: |
          Interrupt for signaling when conversion results exceed the high limit
          for ADC readings or fall below the low limit for them. Interrupt
          source must be attached to COUT pin.
      - description: |
          Alert output that asserts low during a number of different error
          conditions. The interrupt source must be attached to FAULT pin.
```

machschmitt commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,254 @@
+#include <linux/module.h>
+#include <linux/spi/spi.h>
+#include <linux/iio/iio.h>
+#include <linux/of.h>
```
including open firmware header is probably not needed for now.
```diff
-#include <linux/of.h>
```

machschmitt reviewed (commented):
Hi @MarileneGarcia,

This initial driver version looks very good.
I'm leaving a few suggestions.
Aside from a few potentially buggy things, most of the comments are about code style nitpicking.

My advise is to address those last nits and send it to the mailing lists. Experienced kernel maintainers may find further possible improvements for this driver but I do think this is a very good starting point for upstream review.

MarileneGarcia committed:
dt-bindings: iio: adc: Add MAX14001

Add device-tree documentation for MAX14001/MAX14002 ADCs.
The MAX14001/MAX14002 are isolated, single-channel analog-to-digital
converters with programmable voltage comparators and inrush current
control optimized for configurable binary input applications.

Signed-off-by: Marilene Andrade Garcia <marilene.agarcia@gmail.com>

MarileneGarcia committed:
iio: adc: Add basic support for MAX14001

The MAX14001/MAX14002 are configurable, isolated 10-bit ADCs for
multi-range binary inputs. Besides the ADC readings, the MAX14001/MAX14002
offers more features, like a binary comparator, a filtered reading that
can provide the average of the last 2, 4, or 8 ADC readings, and an inrush
comparator that triggers the inrush current. There is also a fault feature
that can diagnose seven possible fault conditions. And an option to select
an external or internal ADC voltage reference.

Add basic support for MAX14001/MAX14002 with the following features:
- Raw ADC reading.
- Filtered ADC average reading with the default configuration.

Signed-off-by: Marilene Andrade Garcia <marilene.agarcia@gmail.com>

MarileneGarcia committed:
overlays: rpi-max14001-pmb-overlay.dts: Add MAX14001PMB example overlay

The MAX14001 is a configurable, isolated 10-bit ADC for multi-range binary
inputs. Besides the ADC readings, the MAX14001 offers more features, like
a binary comparator, a filtered reading that can provide the average of
the last 2, 4, or 8 ADC readings, and an inrush comparator that triggers
the inrush current. There is also a fault feature that can diagnose seven
possible fault conditions. And an option to select an external or internal
ADC voltage reference.

The MAX14001PMB provides the hardware to evaluate the MAX14001 to measure
two channels of data, line voltage and load current.

Signed-off-by: Marilene Andrade Garcia <marilene.agarcia@gmail.com>

machschmitt commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,215 @@
+		{
+			.tx_buf = &st->tx_buffer,
+			.len = sizeof(st->tx_buffer),
+			.bits_per_word = 16,
```
I have just recalled one important detail about using higher bits_per_word settings. The SPI subsystem will store data in native CPU endianness in memory. When using 8 bits_per_word, the data gets pushed to memory in device endianness. Though, since the buffers are `__be16`, they will be accessed in big-endian format. I think ´be16_to_cpu()` might make it work without the need to use 16 bits_per_word.
```c
	st->rx_buffer = bitrev16(be16_to_cpu(st->rx_buffer));
```
If not, then go ahead with bits_per_word = 16. Endianness issues are tricky, and I don't think this driver should be delayed because of that.

machschmitt commented:
> Okay, I believe I’ve addressed them, but I’ll wait for your approval before sending the patches to the maintainers.

The patches look great. But, unfortunately, they don't apply to IIO testing branch. The patches need to be ported to the testing branch of IIO tree. It should not be hard to port the patches by editing the git hunks.
Change patch1 hunk to be like
```
@@ -14952,6 +14952,13 @@ F:	drivers/iio/amplifiers/hmc425a.c
 F:	Documentation/devicetree/bindings/iio/adc/maxim,max11205.yaml
 F:	drivers/iio/adc/max11205.c
 
+ANALOG DEVICES INC MAX14001 DRIVER
+M:	Marilene Andrade Garcia <marilene.agarcia@gmail.com>
+L:	linux-iio@vger.kernel.org
+S:	Maintained
+W:	https://ez.analog.com/linux-software-drivers
+F:	Documentation/devicetree/bindings/iio/adc/adi,max14001.yaml
+
 MAXIM MAX17040 FAMILY FUEL GAUGE DRIVERS
 R:	Iskren Chernev <iskren.chernev@gmail.com>
 R:	Krzysztof Kozlowski <krzk@kernel.org>
```
change patch2 to be like
```
@@ -14958,6 +14958,7 @@ L:	linux-iio@vger.kernel.org
 S:	Maintained
 W:	https://ez.analog.com/linux-software-drivers
 F:	Documentation/devicetree/bindings/iio/adc/adi,max14001.yaml
+F:	drivers/iio/adc/max14001.c
 
 MAXIM MAX17040 FAMILY FUEL GAUGE DRIVERS
 R:	Iskren Chernev <iskren.chernev@gmail.com>
diff --git a/drivers/iio/adc/Kconfig b/drivers/iio/adc/Kconfig
index d3d12ed064e5..22b348bb675a 100644
--- a/drivers/iio/adc/Kconfig
+++ b/drivers/iio/adc/Kconfig
@@ -979,6 +979,16 @@ config MAX11205
 	  To compile this driver as a module, choose M here: the module will be
 	  called max1363.
 
+config MAX14001
+	tristate "Analog Devices MAX14001/MAX14002 ADCs driver"
+	depends on SPI
+	help
+	  Say yes here to build support for Analog Devices MAX14001/MAX14002
+	  Configurable, Isolated 10-bit ADCs for Multi-Range Binary Inputs.
+
+	  To compile this driver as a module, choose M here: the module will be
+	  called max14001.
+
 config MAX34408
 	tristate "Maxim max34408/max344089 ADC driver"
 	depends on I2C
diff --git a/drivers/iio/adc/Makefile b/drivers/iio/adc/Makefile
index 6a71631c9842..bc4c6d7a57dd 100644
--- a/drivers/iio/adc/Makefile
+++ b/drivers/iio/adc/Makefile
@@ -86,6 +86,7 @@ obj-$(CONFIG_MAX11205) += max11205.o
 obj-$(CONFIG_MAX11410) += max11410.o
 obj-$(CONFIG_MAX1241) += max1241.o
 obj-$(CONFIG_MAX1363) += max1363.o
+obj-$(CONFIG_MAX14001) += max14001.o
 obj-$(CONFIG_MAX34408) += max34408.o
 obj-$(CONFIG_MAX77541_ADC) += max77541-adc.o
 obj-$(CONFIG_MAX9611) += max9611.o
```
We may talk more about it at next week's meeting.
